newIDE/app/src/EventsSheet/index.js

```diff
@@ -1084,406 +1115,427 @@ export default class EventsSheet extends React.Component<Props, State> {
+                        onClose={() =>
+                          this.setState({
+                            serializedEventsToExtract: null,
+                          })
+                        }
+                        serializedEvents={this.state.serializedEventsToExtract}
+                        onCreate={(extensionName, eventsFunction) => {
+                          this.props.onCreateEventsFunction(
+                            extensionName,
+                            eventsFunction
+                          );
+                          this._replaceSelectionByEventsFunction(
+                            extensionName,
+                            eventsFunction
+                          );
+                          this.setState({
+                            serializedEventsToExtract: null,
+                          });
+                        }}
+                      />
+                    )}
+                    {this.state.textEditedEvent && (
+                      <EventTextDialog
+                        event={this.state.textEditedEvent}
+                        onApply={() => {
+                          this.closeEventTextDialog();
+                          this._saveChangesToHistory();
+                        }}
+                        onClose={this.closeEventTextDialog}
+                      />
+                    )}
+                    <InfoBar
+                      identifier="edit-instruction-explanation"
+                      message={
+                        <Trans>
+                          Double click on a condition or action to edit it.
+                        </Trans>
+                      }
+                      touchScreenMessage={
+                        <Trans>
+                          Double tap a condition or action to edit it. Long
+                          press to show more options.
+                        </Trans>
+                      }
+                      show={hasInstructionSelected(this.state.selection)}
+                    />
+                  </div>
                 )}
-              </PreferencesContext.Consumer>
+              </EventsSearcher>
             )}
           </ResponsiveWindowMeasurer>
         )}
       </ScreenTypeMeasurer>
     );
   }
 }
+
+export type EventsSheetInterface = {|
+  updateToolbar: () => void,
+|};
+
+// EventsSheet is a wrapper so that the component can use multiple
+// context in class methods while correctly exposing the interface.
+const EventsSheet = (props, ref) => {
+  React.useImperativeHandle(ref, () => ({
+    updateToolbar,
+  }));
+
+  const component = React.useRef<?EventsSheetComponentWithoutHandle>(null);
+  const updateToolbar = () => {
+    console.log(component);
+    if (component.current) component.current.updateToolbar();
+  };
+  return (
+    <AuthenticatedUserContext.Consumer>
+      {(authenticatedUser: AuthenticatedUser) => (
+        <PreferencesContext.Consumer>
+          {userPreferences => (
```

Comment:
nitpicking: Use `useContext` as you're in a functional component - and say good bye to these ugly indentation of render props :D 

It's a good first step toward converting one day the entirety of EventsSheet into a functional component.